hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/schemas/InitialModServiceTokenSchema.java

```diff
@@ -184,8 +189,44 @@ public void setStakingFs(
         this.mnc = mnc;
     }
 
+    /**
+     * Updates in-state staking info to match the address book.
+     * <ol>
+     *     <li>For any node with staking info in state that is no longer in the address book,
+     *     marks it deleted and sets its weight to zero.</li>
+     *     <li>For any node in the address book that is not in state, initializes its staking info.</li>
+     * </ol>
+     *
+     * @param ctx {@link MigrationContext} for this schema restart operation
+     */
     @Override
-    public void restart(@NonNull MigrationContext ctx) {}
+    public void restart(@NonNull MigrationContext ctx) {
+        // We need to validate and mark any node that are removed during upgrade as deleted.
+        // Since restart is called in the schema after an upgrade, and we don't want to depend on
+        // schema version change, validate all the nodeIds from the addressBook in state and mark
+        // them as deleted if they are not yet deleted in staking info.
+        final var stakingToState = ctx.newStates().<EntityNumber, StakingNodeInfo>get(STAKING_INFO_KEY);
```

Comment:
I think all schemas will be registered by this point. So should not be an issue 🤔